### PR TITLE
Implement recursive package coverage

### PR DESCRIPTION
**Pull request by lowjoel**
As requested in https://youtrack.jetbrains.com/issue/GO-3620 by @zolotov, this implements support for recursive package coverage with line annotations.

This uses the package-coverage tool: https://github.com/corsc/go-tools/tree/master/package-coverage

### CODE REVIEW

**Comment by zolotov** [comment]
The solution is good, thank you! Although implementation and UX might be improved.

1. TestRunningState is not the best place for merging coverage report and looking for special executable. It would be better to have this functionality closer to `GoCoverageRunner` and `GoCoverageProgramRunner`
2. User will never know how to run coverage on a directory. It would be better if GoCoverageRunner could recognize running on directory and suggest to download and install missing package and make sure its executable is in `PATH`.
3. How good does this work on windows?

As a solution of the second bullet, we can actually bundle `package-coverage` executable.

**Comment by zolotov** [comment]
Also, if you're going to continue on this PR, please sign https://www.jetbrains.com/agreements/cla/ first. Thank you.

**Comment by lowjoel** [comment]
Hi @zolotov yep, I've signed the CLA.

I'll try to refactor the code to solve Point 1; I think for Point 2 I'd need some guidance from you because that process is pretty unfamiliar to me. We can talk over Gitter if that's good for you.

I'm trying everything out on a Windows machine right now, I'll post updates here.

**Comment by lowjoel** [comment]
@zolotov: I've done some refactoring to the code, hope this is better :)

This works quite well on Windows, once the package-coverage binary is in place.

**Comment by lowjoel** [comment]
One more change to display a notification when the `package-coverage` binary isn't installed.

File: src/com/goide/runconfig/testing/coverage/GoCoverageMerger.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        writer.newLine();
+        mergeCoverage(writer, new File(fromPath));
+      }
+    } catch (IOException e) {
+      LOG.error(e.getMessage());
+    }
+  }
+
+  private static void mergeCoverage(BufferedWriter writer, File directory) throws IOException {
+    for (File file : directory.listFiles()) {
```
Comment (zolotov) [inline src/com/goide/runconfig/testing/coverage/GoCoverageMerger.java]: using VirtualFile will be much-much faster. See also VfsUtil class

File: src/com/goide/runconfig/testing/coverage/GoCoverageMerger.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+      if (file.isDirectory()) {
+        mergeCoverage(writer, file);
+      } else if ("profile.cov".equals(file.getName())) {
```
Comment (zolotov) [inline src/com/goide/runconfig/testing/coverage/GoCoverageMerger.java]: code style

File: src/com/goide/runconfig/testing/coverage/GoCoverageProgramRunner.java [hunk 3]
```diff
@@ -32,14 +34,77 @@
...
+
+      NotificationGroup group = new NotificationGroup(
+          ID, NotificationDisplayType.BALLOON, true
+      );
+      Notification notification = group.createNotification(
+          "Recursive Directory Coverage",
+          "Directory coverage can be computed recursively if the <code>corsc/go-tools/package-coverage</code> package is installed on this computer",
+          NotificationType.INFORMATION,
+          null);
+      Notifications.Bus.notify(notification, null);
```
Comment (zolotov) [inline src/com/goide/runconfig/testing/coverage/GoCoverageProgramRunner.java]: see GoDownloadableFileAction, there user can invoke `go get` right from notification

File: src/com/goide/runconfig/testing/coverage/GoCoverageProgramRunner.java [hunk 4]
```diff
@@ -32,14 +34,77 @@
...
+          myConfiguration.getConfigurationModule().getModule());
+    }
+
+    @Override
+    public boolean canRecursiveCoverage() {
+      if (hasPackageCoverageExecutable()) {
+        return true;
+      }
+
+      NotificationGroup group = new NotificationGroup(
```
Comment (zolotov) [inline src/com/goide/runconfig/testing/coverage/GoCoverageProgramRunner.java]: See GoConstants.GO_NOTIFICATION_GROUP

File: src/com/goide/runconfig/testing/coverage/GoCoverageProgramRunner.java [hunk 5]
```diff
@@ -32,14 +34,77 @@
...
+          ID, NotificationDisplayType.BALLOON, true
+      );
+      Notification notification = group.createNotification(
```
Comment (zolotov) [inline src/com/goide/runconfig/testing/coverage/GoCoverageProgramRunner.java]: Usually, it's quite unexpected that methods like `canRectursiveCoverage` besides giving the answer do also some other things, moreover in UI. I'd say that showing notification should be on client, who creates this patcher

**Comment by zolotov** [comment]
Thank you, looks much better. Please make some updates in PR and I'll happily merge it

File: src/com/goide/runconfig/testing/coverage/GoCoverageMerger.java (hunk 1 above)
Comment (lowjoel) [inline src/com/goide/runconfig/testing/coverage/GoCoverageMerger.java]: Using VirtualFile doesn't seem to give me an up-to-date view of the file, the resulting coverage generated is empty. Ideas?

File: src/com/goide/runconfig/testing/coverage/GoCoverageProgramRunner.java (hunk 3 above)
Comment (zolotov) [inline src/com/goide/runconfig/testing/coverage/GoCoverageProgramRunner.java]: You should use it as an example

File: src/com/goide/runconfig/testing/coverage/GoCoverageMerger.java (hunk 1 above)
Comment (zolotov) [inline src/com/goide/runconfig/testing/coverage/GoCoverageMerger.java]: You should ask VFS to reload external FS changes, e.g. via `VfsUtil.markDirtyAndRefresh`.

**Comment by lowjoel** [comment]
@zolotov I've tried my best to address your comments, have a look at one last question in the review.

File: src/com/goide/runconfig/testing/coverage/GoCoverageMerger.java (hunk 1 above)
Comment (lowjoel) [inline src/com/goide/runconfig/testing/coverage/GoCoverageMerger.java]: Occasionally I get this:
```
Assertion failed: Write access is allowed inside write-action only (see com.intellij.openapi.application.Application.runWriteAction())
java.lang.Throwable
	at com.intellij.openapi.diagnostic.Logger.assertTrue(Logger.java:168)
	at com.intellij.openapi.application.impl.ApplicationImpl.assertWriteAccessAllowed(ApplicationImpl.java:1232)
	at com.intellij.openapi.vfs.newvfs.persistent.PersistentFSImpl.processEvents(PersistentFSImpl.java:736)
	at com.intellij.openapi.vfs.newvfs.persistent.PersistentFSImpl.processEvent(PersistentFSImpl.java:672)
	at com.intellij.openapi.vfs.newvfs.persistent.PersistentFSImpl.deleteFile(PersistentFSImpl.java:452)
	at com.intellij.openapi.vfs.newvfs.impl.VirtualFileSystemEntry.delete(VirtualFileSystemEntry.java:204)
	at com.goide.runconfig.testing.coverage.GoCoverageMerger.lambda$mergeCoverage$0(GoCoverageMerger.java:44)
	at com.intellij.openapi.vfs.VfsUtilCore$1.visitFileEx(VfsUtilCore.java:256)
	at com.intellij.openapi.vfs.VfsUtilCore.visitChildrenRecursively(VfsUtilCore.java:271)
	at com.intellij.openapi.vfs.VfsUtilCore.visitChildrenRecursively(VfsUtilCore.java:303)
	at com.intellij.openapi.vfs.VfsUtilCore.visitChildrenRecursively(VfsUtilCore.java:303)
	at com.intellij.openapi.vfs.VfsUtilCore.visitChildrenRecursively(VfsUtilCore.java:303)
	at com.intellij.openapi.vfs.VfsUtilCore.visitChildrenRecursively(VfsUtilCore.java:303)
	at com.intellij.openapi.vfs.VfsUtilCore.visitChildrenRecursively(VfsUtilCore.java:303)
	at com.intellij.openapi.vfs.VfsUtilCore.iterateChildrenRecursively(VfsUtilCore.java:251)
	at com.goide.runconfig.testing.coverage.GoCoverageMerger.mergeCoverage(GoCoverageMerger.java:29)
	at com.goide.runconfig.testing.coverage.GoCoverageMerger.MergeCoverage(GoCoverageMerger.java:22)
	at com.goide.runconfig.testing.coverage.GoCoverageProcessListener.processTerminated(GoCoverageProcessListener.java:27)
```

Not sure what I can do about this.

File: src/com/goide/runconfig/testing/coverage/GoCoverageMerger.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+import com.intellij.openapi.vfs.LocalFileSystem;
+import com.intellij.openapi.vfs.VfsUtil;
+import com.intellij.openapi.vfs.VirtualFile;
+
+import java.io.*;
+
+public class GoCoverageMerger {
+  protected static final Logger LOG = Logger.getInstance("#com.goide.runconfig.testing.coverage.GoCoverageMerger");
+
+  public static void MergeCoverage(String fromPath, String toFile) {
```
Comment (zolotov) [inline src/com/goide/runconfig/testing/coverage/GoCoverageMerger.java]: Code style, in java method names shoudn't be capitalized

File: src/com/goide/runconfig/testing/coverage/GoCoverageMerger.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    try (
+        FileWriter stream = new FileWriter(toFile, false);
+        BufferedWriter writer = new BufferedWriter(stream)
+    ) {
+      writer.write("mode: set");
+      writer.newLine();
+      VirtualFile fromFile = LocalFileSystem.getInstance().findFileByPath(fromPath);
+      VfsUtil.markDirtyAndRefresh(false, true, true, fromFile);
```
Comment (zolotov) [inline src/com/goide/runconfig/testing/coverage/GoCoverageMerger.java]: Merger should not know anything about VFS refreshing, it should just merge – that's it. 
It's a client who knows that the files were appeared a moment ago and need to be refreshed.

File: src/com/goide/runconfig/testing/coverage/GoCoverageMerger.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    } catch (IOException e) {
+      LOG.error(e.getMessage());
+    }
+  }
+
+  private static void mergeCoverage(BufferedWriter writer, VirtualFile directory) {
+    VfsUtil.iterateChildrenRecursively(directory,
+        null,
+        virtualFile -> {
+          if (!"profile.cov".equals(virtualFile.getName())) {
```
Comment (zolotov) [inline src/com/goide/runconfig/testing/coverage/GoCoverageMerger.java]: this is what `filter` parameter of `iterateChildrenRecursively` is needed for

File: src/com/goide/runconfig/testing/coverage/GoCoverageMerger.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+                BufferedReader reader = new BufferedReader(streamReader)
+            ) {
+              copyCoverage(writer, reader);
+            }
+            virtualFile.delete(null);
+            return true;
+          }
+          catch (IOException e) {
+            LOG.error(e.getMessage());
+            return false;
```
Comment (zolotov) [inline src/com/goide/runconfig/testing/coverage/GoCoverageMerger.java]: do you really need to stop processing if you failed to read a single file?

File: src/com/goide/runconfig/testing/coverage/GoCoverageMerger.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+  private static void mergeCoverage(BufferedWriter writer, VirtualFile directory) {
+    VfsUtil.iterateChildrenRecursively(directory,
+        null,
+        virtualFile -> {
+          if (!"profile.cov".equals(virtualFile.getName())) {
+            return true;
+          }
+
+          try {
+            try (
```
Comment (zolotov) [inline src/com/goide/runconfig/testing/coverage/GoCoverageMerger.java]: com.intellij.openapi.util.io.StreamUtil#copyStreamContent

File: src/com/goide/runconfig/testing/coverage/GoCoverageProcessListener.java [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+  @Override
+  public void startNotified(ProcessEvent event) {}
+
+  @Override
+  public void processTerminated(ProcessEvent event) {
+    if (event.getExitCode() != 0 || new File(myCoverageFilePath).exists()) {
+      return;
+    }
+
+    GoCoverageMerger.MergeCoverage(myCoverageDir, myCoverageFilePath);
```
Comment (zolotov) [inline src/com/goide/runconfig/testing/coverage/GoCoverageProcessListener.java]: Please make it OOP. Merger that just contain some static classes is not Merger but some Utility class.

File: src/com/goide/runconfig/testing/coverage/GoCoverageProgramRunner.java [hunk 12]
```diff
@@ -68,13 +124,46 @@ protected RunContentDescriptor doExecute(@NotNull RunProfileState state, @NotNul
...
+        !runningStatePatcher.hasPackageCoverageExecutable()) {
+      promptForPackageCoverageTool(runConfiguration);
+    }
 
     ExecutionResult executionResult = state.execute(environment.getExecutor(), this);
     if (executionResult == null) {
       return null;
     }
+    executionResult.getProcessHandler().addProcessListener(new GoCoverageProcessListener(
+        runConfiguration.getDirectoryPath(),
```
Comment (zolotov) [inline src/com/goide/runconfig/testing/coverage/GoCoverageProgramRunner.java]: So all reports are generated right in user's sources?
Sorry, this is not acceptable.

1. IDE must not pollute user's sources files with other unrelated files
2. IDE should not overwrite already existing files by invoking Coverage
3. Recursing iterating across all directory files while coverage merging will be too long because of iterating the files that have nothing to do with coverage

Comment (lowjoel) [inline src/com/goide/runconfig/testing/coverage/GoCoverageProgramRunner.java]: Then I see no other way of doing this. If this is a deal breaker I'll close the PR.

**Comment by lowjoel** [comment]
@zolotov let me know if the final point is a deal breaker. If it is, I'll close it and will wait for you all to implement coverage.

File: src/com/goide/runconfig/testing/coverage/GoCoverageProgramRunner.java (hunk 12 above)
Comment (zolotov) [inline src/com/goide/runconfig/testing/coverage/GoCoverageProgramRunner.java]: Yes, each of items is important.

> The only way to do this with your points considered will be for the IDE

It's not the only way. I believe `package-coverage` is not the only tool for multi-package coverage. The PR can be applied to any of such tool. These packages I found for couple minutes:

https://github.com/go-playground/overalls
https://github.com/haya14busa/goverage
https://github.com/pierrre/gotestcover
https://github.com/axw/gocov

Also, `package-coverage` can be changed as well to allow customize output directory.

Comment (lowjoel) [inline src/com/goide/runconfig/testing/coverage/GoCoverageProgramRunner.java]: https://github.com/go-playground/overalls doesn't satisfy (1)
https://github.com/pierrre/gotestcover is deprecated
https://github.com/axw/gocov writes to JSON

I'll see if https://github.com/haya14busa/goverage can be adapted to it. I don't own package-coverage, so it's not something I can change easily either. The set of tools that satisfies all 3 points _and_ can be almost compatible with the `go test` command line is small.

Comment (zolotov) [inline src/com/goide/runconfig/testing/coverage/GoCoverageProgramRunner.java]: > I'll see if https://github.com/haya14busa/goverage can be adapted to it. 

Thanks.

> I don't own package-coverage, so it's not something I can change easily either.

Well, you can change Go Plugin for IntelliJ for sure, why do not change package-coverage? ;) 
I think changing `package-coverage` is easier than reimplementing IDE to run a process for each package in IDE.

Anyway, you did a good work and I assure you it won't be thrown away. Even if you close the PR, I'll grab these changes eventually.

Comment (lowjoel) [inline src/com/goide/runconfig/testing/coverage/GoCoverageProgramRunner.java]: > Well, you can change Go Plugin for IntelliJ for sure, why do not change package-coverage? ;)
I think changing package-coverage is easier than reimplementing IDE to run a process for each package in IDE.

Silly me. That _does_ seem easier. However goverage seems to do that already so it's probably better to just give that a shot. It's a relatively new project (a few weeks before I started writing these changes) so 🤞

**Comment by lowjoel** [comment]
@zolotov: strangely after running the IDE today, subsequent builds fail with `Cannot add a configuration with name 'idea' as a configuration with that name already exists.` I've already cleared my Gradle caches... any idea what's up? Google doesn't come up with anything useful.

**Comment by lowjoel** [comment]
OK, upgrading the Gradle Intellij plugin seems to work.

**Comment by lowjoel** [comment]
@zolotov: I've removed the process listener and the coverage merger, since goverage handles that internally. How does this look now?

**Comment by zolotov** [comment]
Looks good so far. I need to take a look how good it work and will merge it after that. Thank you!

File: src/com/goide/runconfig/testing/coverage/GoCoverageProgramRunner.java [hunk 13]
```diff
@@ -32,13 +36,57 @@
...
+    }
+
+    @Override
+    public void beforeTarget(@NotNull GoExecutor executor) {
+      if (hasPackageCoverageExecutable()) {
+        executor.withExePath(myPackageCoverageExecutable.getPath());
+        executor.withParameters("-v");
+      }
+      else {
+        super.beforeTarget(executor);
```
Comment (zolotov) [inline src/com/goide/runconfig/testing/coverage/GoCoverageProgramRunner.java]: so if a user doesn't have goverage then you will add GoToolParams twice, here and in `afterTarget`. It leads to this:

image

File: src/com/goide/runconfig/testing/coverage/GoCoverageProgramRunner.java [hunk 14]
```diff
@@ -68,7 +116,13 @@ protected RunContentDescriptor doExecute(@NotNull RunProfileState state, @NotNul
     }
     FileDocumentManager.getInstance().saveAllDocuments();
     CoverageEnabledConfiguration coverageEnabledConfiguration = CoverageEnabledConfiguration.getOrCreate(runConfiguration);
-    runningState.setCoverageFilePath(coverageEnabledConfiguration.getCoverageFilePath());
+    CoveragePatcher runningStatePatcher = new CoveragePatcher(coverageEnabledConfiguration, runConfiguration);
+    runningState.setPatcher(runningStatePatcher);
+
+    if (runConfiguration.getKind() == GoTestRunConfiguration.Kind.DIRECTORY &&
+        !runningStatePatcher.hasPackageCoverageExecutable()) {
+      promptForPackageCoverageTool(runConfiguration);
```
Comment (zolotov) [inline src/com/goide/runconfig/testing/coverage/GoCoverageProgramRunner.java]: Why do not interrupt running here?

File: src/com/goide/runconfig/testing/coverage/GoCoverageProgramRunner.java [hunk 15]
```diff
@@ -77,4 +131,25 @@ protected RunContentDescriptor doExecute(@NotNull RunProfileState state, @NotNul
     CoverageHelper.attachToProcess(runConfiguration, executionResult.getProcessHandler(), environment.getRunnerSettings());
     return new RunContentBuilder(executionResult, environment).showRunContent(environment.getContentToReuse());
   }
+
+  private void promptForPackageCoverageTool(@NotNull final GoTestRunConfiguration configuration) {
+    Notification notification = GoConstants.GO_NOTIFICATION_GROUP.createNotification(
+        "Recursive Directory Coverage",
+        "Directory coverage can be computed recursively if the <code>" + GO_GET_BINARY_NAME + "</code> package is installed.<br><br><a href=\"" + GO_GET_BINARY_LINK + "\">Install Package Coverage</a>",
+        NotificationType.INFORMATION,
+        (notification1, hyperlinkEvent) -> {
```
Comment (zolotov) [inline src/com/goide/runconfig/testing/coverage/GoCoverageProgramRunner.java]: Duplicates listener from `GoDownloadableFileAction`

File: src/com/goide/runconfig/testing/coverage/GoCoverageProgramRunner.java [hunk 16]
```diff
@@ -32,13 +36,57 @@
...
+  private static final String GO_GET_BINARY_LINK = "goGetPackageCoverageLink";
+  private static final String GO_GET_BINARY_NAME = "github.com/haya14busa/goverage";
+  private static final String GO_GET_BINARY = "goverage";
+
+  private class CoveragePatcher extends GoTestRunningState.ExecutorPatcher {
+    private GoTestRunConfiguration myConfiguration;
+    private CoverageEnabledConfiguration myCoverageConfiguration;
+    private VirtualFile myPackageCoverageExecutable;
+
+    public CoveragePatcher(CoverageEnabledConfiguration coverageConfiguration, GoTestRunConfiguration configuration) {
```
Comment (zolotov) [inline src/com/goide/runconfig/testing/coverage/GoCoverageProgramRunner.java]: The `patcher` is supposed to patch executor, it should not know anything about TestRunConfiguration

**Comment by lowjoel** [comment]
Sorry, I really didn't have the time to work on this. Since 1.10 will fix this, let's use your implementation in GoLand.

**Comment by zolotov** [comment]
@lowjoel sure. Thanks for an attempt!
